### DISCUSSION TITLE
Issues with Windows Server 2019 install batch files using improperly declared variables.

### DISCUSSION THREAD

**Discussion by Kevin-Herr** [rabbitmq/rabbitmq-server#9888] [Other] [answered]
### Describe the bug

It appears that the Windows installer has batch files with variables declared incorrectly or in a way that Windows Server 2019 will not install without errors. They're using !VARIABLE! instead of %VARIABLE% in a lot of places and I'm just taking a wild guess at this point that this might be why I keep getting the error "Unrecognized option neighbours". 

If this is as designed and it's something else causing my issues, any advice would be appreciated because I'm at a loss after many reboots and hundreds of installation attempts. 

**Example below of a snippet from the installers files in a test batch file.**

C:\>C:\Users\admin\Desktop\test.bat

C:\>if not exist "!ERLANG_HOME!\bin\erl.exe" (
echo.
 echo ******************************
 echo ERLANG_HOME not set correctly.
 echo ******************************
 echo.
 echo Please either set ERLANG_HOME to point to your Erlang installation or place the
 echo RabbitMQ server distribution in the Erlang lib folder.
 echo.
)

`******************************`
ERLANG_HOME not set correctly.
`******************************`

Please either set ERLANG_HOME to point to your Erlang installation or place the
RabbitMQ server distribution in the Erlang lib folder.

**Example below of a snippet from the installers files with an adjusted variable declaration.**

C:\>C:\Users\admin\Desktop\test.bat

C:\>if not exist "C:\Program Files\Erlang OTP\bin\erl.exe" (
echo.
 echo ******************************
 echo ERLANG_HOME not set correctly.
 echo ******************************
 echo.
 echo Please either set ERLANG_HOME to point to your Erlang installation or place the
 echo RabbitMQ server distribution in the Erlang lib folder.
 echo.
)

C:\>

### Reproduction steps

1. On Windows Server 2019, install Erlang (otp_win64_25.3.2.7.exe).
2. Set ERLANG_HOME environment variable.
3. Install RabbitMQ (rabbitmq-server-3.12.8.exe).
...

### Expected behavior

A smooth installation without receiving the error at the bottom of the output below.

Output folder: C:\Program Files\RabbitMQ Server\rabbitmq_server-3.12.8\plugins
Output folder: C:\Users\admin\AppData\Roaming\RabbitMQ
Created uninstaller: C:\Program Files\RabbitMQ Server\uninstall.exe
Installing RabbitMQ service...
C:\Program Files\Erlang OTP\erts-13.2.2.4\bin\erlsrv: Service RabbitMQ added to system.
C:\Program Files\Erlang OTP\erts-13.2.2.4\bin\erlsrv: Unrecognized option neighbours:.

**Comment by michaelklishin** [comment]
We don't have enough information to conclude much. This may be PowerShell-version specific and there are no details about that, nor there are any details about the value of the `ERLANG_HOME` environment variable.

Are you sure you [only have one Erlang version installed](https://rabbitmq.com/windows-quirks.html)? Why do you need to explicitly set `ERLANG_HOME` and why use 25.3 over the latest?

I do not recall other reports like this. @lukebakken do you?

**Reply by lukebakken** [reply]
No, this must be specific to @Kevin-Herr's environment.

**Reply by Kevin-Herr** [reply]
I've attempted many combinations of supported versions, I just came here with the most recent one I've attempted and there is only one Erlang version installed. This has been a problem on two newly built servers we are preparing for a cluster.

**Reply by Kevin-Herr** [reply]
Attempted to install with Chocolatey on the other server and encountered the same issue and error. The pop up mentioned was the same one as when I was attempting to manually install with the executables. 

![image](https://github.com/rabbitmq/rabbitmq-server/assets/37375406/4c72a5be-480a-44e2-b4d8-d75c7e678664)

**Reply by Kevin-Herr** [reply]
**From chocolatey.log**

2023-11-07 12:50:47,177 4748 [WARN ] - 
Chocolatey installed 1/2 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
2023-11-07 12:50:47,177 4748 [INFO ] - 
2023-11-07 12:50:47,193 4748 [ERROR] - Failures
2023-11-07 12:50:47,193 4748 [ERROR] -  - rabbitmq (exited 2) - Error while running 'C:\ProgramData\chocolatey\lib\rabbitmq\tools\chocolateyInstall.ps1'.
 See log for details.
2023-11-07 12:50:47,193 4748 [DEBUG] - Sending message 'PostRunMessage' out if there are subscribers...
2023-11-07 12:50:47,193 4748 [DEBUG] - Exiting with 2

**Comment by michaelklishin** [comment]
Our package tests use Windows 2016 and 2022, and report `ERLANG_HOME` being set using

``` batch
set ERLANG_HOME=%ProgramFiles%\Erlang OTP
```

I could not find anything about the currently used environment variable syntax being removed starting with Windows Server 2019 or PowerShell 5.1 (which seems to be the default in Windows Server 2019).

Combined with the fact that this is not something commonly reported and if it was an issue it would likely break for the majority of Windows users, this must be something system-specific.

**Reply by Kevin-Herr** [reply]
It's not an issue with the setting of the variable, it's the calling of the variable in the scripts inside. It's attempting to call them with !VARIABLE! instead of %VARIABLE% which was what I was trying to point out in the examples.

**Reply by lukebakken** [reply]
`!VARIABLE!` is delayed expansion, which works fine on every Windows system I've tested the installers on. There must be something different about your systems 🤷‍♂️

**Comment by lukebakken** [comment]
@Kevin-Herr would you do the following with a clean Windows Server 2019 environment -
* Start an administrative `cmd.exe` process.
* Run the `set` and `ver` commands, capture the output to a file, and _attach_ the file to your response.
* Install Erlang via the exe installer.
* Do **NOT** set `ERLANG_HOME`
* Install RabbitMQ via the exe installer.

**Reply by Kevin-Herr** [reply]
Hello,

I attempted to do that when I had started but setting the ERLANG_HOME variable came after the many failed attempts at installation.

C:\>set
ALLUSERSPROFILE=C:\ProgramData
APPDATA=C:\Users\admin\AppData\Roaming
CLIENTNAME=IT-LT2
CommonProgramFiles=C:\Program Files\Common Files
CommonProgramFiles(x86)=C:\Program Files (x86)\Common Files
CommonProgramW6432=C:\Program Files\Common Files
COMPUTERNAME=RABBITMQ-01
ComSpec=C:\Windows\system32\cmd.exe
DriverData=C:\Windows\System32\Drivers\DriverData
ERLANG_HOME=C:\Program Files\Erlang OTP
HOMEDRIVE=H:
HOMEPATH=\
HOMESHARE=\\fileserver\Homes\Admin
LOCALAPPDATA=C:\Users\admin\AppData\Local
LOGONSERVER=\\DC-04
NUMBER_OF_PROCESSORS=8
OS=Windows_NT
Path=C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Erlang OTP\bin;C:\Users\admin\AppData\Local\Microsoft\WindowsApps;
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
PROCESSOR_ARCHITECTURE=AMD64
PROCESSOR_IDENTIFIER=Intel64 Family 6 Model 63 Stepping 0, GenuineIntel
PROCESSOR_LEVEL=6
PROCESSOR_REVISION=3f00
ProgramData=C:\ProgramData
ProgramFiles=C:\Program Files
ProgramFiles(x86)=C:\Program Files (x86)
ProgramW6432=C:\Program Files
PROMPT=$P$G
PSModulePath=C:\Program Files\WindowsPowerShell\Modules;C:\Windows\system32\WindowsPowerShell\v1.0\Modules
PUBLIC=C:\Users\Public
SESSIONNAME=RDP-Tcp#0
SystemDrive=C:
SystemRoot=C:\Windows
TEMP=C:\Users\admin\AppData\Local\Temp\2
TMP=C:\Users\admin\AppData\Local\Temp\2
USERDNSDOMAIN=DOMAIN.ORG
USERDOMAIN=DOMAIN
USERDOMAIN_ROAMINGPROFILE=DOMAIN
USERNAME=admin
USERPROFILE=C:\Users\admin
windir=C:\Windows


C:\>ver

Microsoft Windows [Version 10.0.17763.4974]

**Comment by Kevin-Herr** [comment]
Solved. 

Issue was not variables mentioned previously but a single variable, RABBITMQ_BASE, and the software for some reason not being able to use a base folder if it was created by the installer. 

Manually created the following file structures and ensured that SYSTEM and Administrators were granted full access and that there were no spaces in the paths.

>E:\Applications\Erlang\
>E:\Applications\RabbitMQ\
>E:\RabbitMQBASE\

Set the global variable for RABBITMQ_BASE to "E:\RabbitMQBASE". **This did not seem to be optional.**

**Command**: `setx RABBITMQ_BASE "E:\RabbitMQBASE" /M`

Made sure to delete any phantom RabbitMQ services from failed installations. **This may have been optional.**

**Command**: `sc delete RabbitMQ`

Installed Erlang to the directory for it mentioned above without spaces. **This did not seem to be optional.**

Installed RabbitMQ to the directory for it mentioned above without spaces. **Unknown if a directory without spaces was optional for the RabbitMQ installer.**

**Reply by lukebakken** [reply]
Thank you very much for following up.

Erlang and RabbitMQ can be installed to directories that contain spaces. By default, both install to `C:\Program Files`. This is the usual practice on Windows.

[`RABBITMQ_BASE`](https://www.rabbitmq.com/relocate.html#environment-variables) is an optional variable that can be used to move all of the on-disk folders for RabbitMQ as you have found. Otherwise, they install to the `%APPDATA%` directory of the admin account used to install RabbitMQ. Yes, this is weird and we eventually plan to use `C:\ProgramData` but other RabbitMQ work is higher priority.

**Reply by Kevin-Herr** [reply, marked as answer]
There were other issues as well, adding my notes below in case anyone else runs into them. 

If you were installing the software and it failed, you likely need to do some manual cleanup because the system will believe that some services may exist and were setup properly that were not. The instructions below go through cleanup and manual deployment of the service if you don't want to just reinstall the software.

For installation on an **Active Directory Domain** System, you need to login as a **Local Admin Account** on the server. If you are on a domain account with **Home Drives** declared in Active Directory they are **not able to be manually redirected** and will not change regardless of attempts to. Mine was attempting to use a mapped network drive which is **inaccessible from batch files**.

Apologies for my markup, this platform is new to me and this seemed the cleanest way to organize everything.

**Admin Command Prompt Instructions.**
>Delete orphaned services from failed installations.
	`sc delete RabbitMQ`

>Cleanup failed erlang installation registry keys. Have not tested this command but the key is the correct one to remove.
	`reg delete "HKEY_LOCAL_MACHINE\SOFTWARE\Ericsson\Erlang\ErlSrv\1.1\RabbitMQ" /F`

>Creation / Updating Global Environment Variables
	`setx ERLANG_HOME "E:\Applications\Erlang" /M`
	`setx RABBITMQ_BASE "E:\RabbitMQBASE" /M`

>Creation of directories.
	`mkdir E:\Applications\`
	`mkdir E:\Applications\RabbitMQ\`
	`mkdir E:\Applications\Erlang\`
	`mkdir E:\RabbitMQBASE\`

>Copying the Erlang cookie to your %HOMEDRIVE%. Have not tested this command but the cookie was able to be manually copied. May need to grant your account access to the config folder prior to attempting to copy. Note, my default %HOMEDRIVE% for my admin account was "C:" which is why I'm putting it at the root of C:\.
	`copy "C:\Windows\System32\config\systemprofile\.erlang.cookie" "C:\.erlang.cookie"`

**From here you may be able to just reinstall using the RabbitMQ executable but I did not test that way or continue below.**

>Navigating to the proper directory, in my case a different drive, and running the commands to enable management and install the reinstall the Erlang and Windows Services.
	`E:\`
	`cd E:\Applications\RabbitMQ\rabbitmq_server-3.12.8\sbin`
	`rabbitmq-plugins.bat enable rabbitmq_management`
	`rabbitmq-service.bat install`

>Management access.
	`http://server:15672/`
